framework/kms-keyring.md

```diff
@@ -160,9 +160,10 @@ If the input [encryption materials](#structures.md#encryption-materials) do not
 and this keyring has a [generator](#generator) defined, and onEncrypt MUST attempt to generate a new plaintext data key
 and encrypt that data key by calling [KMS GenerateDataKey](#kms-generatedatakey).
 
-The [KMS client](#kms-client) that calls [KMS GenerateDataKey](#kms-generatedatakey) MUST be the
-client returned by the [client supplier](#client-supplier).
-The client MUST be a client that calls the AWS region specified in the [generator](#generator) ARN.
+If an AWS region can be extracted from the [generator](#generator), then the [KMS client](#kms-client) that calls
+[KMS GenerateDataKey](#kms-generatedatakey) MUST be the client returned by the [client supplier](#client-supplier)
+when given that region as input. If an AWS region cannot be extraced from the [generator](#generator),
+then the KMS Keyring MAY choose an arbitrary region.
```

Comment:
We might want to clarify what we mean by `May choose an arbitrary region`. What we really mean is that if you can't get the region from the key, then the KMS keyring is allowed to use a KMS client from any arbitrary region to make the call to KMS. Maybe just say "If the AWS region cannot be extracted...the KMS Keyring MAY use a KMS client from any region"

Reply:
This is where I hit a bit of a sticking point: Who is the actor here?

# Region discernible from key ID

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring passes the region name to the client supplier.
1. The client supplier provides the KMS Keyring with a client.
1. The KMS Keyring uses that client.

# Region not discernible from key ID

## Keyring is the actor

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring fails to determine region from key ID.
1. The KMS Keyring *through some undefined mechanism* obtains a client.

## Client Supplier is the actor

1. The KMS Keyring *through some undefined mechanism* attempts to determine region from key ID.
1. The KMS Keyring fails to determine region from key ID.
1. The KMS Keyring tells the Client Supplier that it doesn't know what the region is.
1. The Client Supplier MAY provide the KMS Keyring with a client.
1. The KMS Keyring uses that client or fails if the Client Supplier fails to provide one.

Reply:
There are two points being brought up here:
1. We should specify that the client supplier is responsible for determining default/"I can't get a region" behavior instead of the keyring. As said above, this makes sense because otherwise we violate the API contract.
2. In the view of a client supplier, it can be a bit tricky what the default behavior SHOULD be. Generally the supplier should make use of the AWS SDK defualt region, however the "limit regions" supplier example above shows how this might not always be the desired behavior. I think more precisely defining the restrictions on client supplier behavior is out of scope for this PR.

As such, I think these are the next steps for this PR:

- Update client supplier interface to say that it MUST support some input (i.e. empty string) to indicate that the region in unknown
- Update L66 to say that if the region can't be determined, then the Keyring MUST input the "I don't know" input to the client supplier, and if the client supplier returns a client the Keyring MUST use that client.

@mattsb42-aws @seebees Thoughts?

Reply:
Good to me.

Reply:
Sounds good!

Reply:
As I read the spec, the Client Supplier is an interface which we let the customer implement. As such, I don't think it makes sense to define default behavior.

Reply:
How about:

> ### Client Supplier
> A function which MUST take as input either a region string (e.g. `us-east-1`), or some value denoting an unknown region, and MAY return a KMS Client which can make the following API calls in the given AWS region:

and

> If an AWS region cannot be extracted from the generator then the KMS Keyring MUST input a value denoting an unknown region. If the client supplier does not provide any client for the given region for this Encrypt call, OnEncrypt MUST skip that particular CMK.

I've opted not to define the value indicating an unknown region, as that seems fairly language specific.

Reply:
`s/which/that/` but otherwise LGTM @MatthewBennington

Reply:
new wording LGTM